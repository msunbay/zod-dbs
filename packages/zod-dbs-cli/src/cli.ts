import { Command, Option, program } from 'commander';
import { generateZodSchemas } from 'zod-dbs';
import { logDebug, toError, ZodDbsProvider } from 'zod-dbs-core';

import type { ZodDbsCliConfig, ZodDbsCliOptions } from './types.js';

import { enableDebugMode, getConfiguration } from './config.js';
import { importProvider } from './provider.js';
import { getArgumentValue, isSilentMode } from './utils/args.js';
import { logAppName, logError, logSetting } from './utils/logger.js';
import { createProgressHandler } from './utils/progress.js';
import { getAppVersion } from './utils/version.js';

export const runCli = async (cliOptions: ZodDbsCliOptions = {}) => {
  enableDebugMode();
  logDebug('Starting zod-dbs CLI with options:', cliOptions);

  const appVersion = cliOptions.appVersion ?? (await getAppVersion());
  const appName = cliOptions.appName || 'zod-dbs';
  const silent = isSilentMode();

  if (!silent) {
    logAppName(`${appName} CLI v${appVersion}`);
  }

  const config = await getConfiguration(cliOptions);

  program.name(appName);
  program.version(appVersion);
  program.description('Generates Zod schemas from database schema.');

  const provider = await loadProvider(cliOptions.overrides?.provider, config);

  if (!cliOptions.overrides?.provider) {
    program.option('-p,--provider <name>', 'Database provider to use');
  }

  program.option(
    '--config-name <name>',
    'Name of configuration file. E.g. "development" will load "zod-dbs-development.ts"'
  );

  // Add provider-specific dynamic options before parsing the full CLI
  addProviderOptions(program, provider);

  const outputOptions = program.optionsGroup('Output options:');

  const parseBooleanOption = (value: string) => {
    if (value.toLowerCase() === 'true') return true;
    if (value.toLowerCase() === 'false') return false;
    throw new Error(`Invalid boolean value: ${value}`);
  };

  outputOptions.option(
    '-o,--output-dir <path>',
    'Output directory for generated schemas'
  );
  outputOptions.addOption(
    new Option(
      '--module-resolution <type>',
      'Module resolution type for generated files. (defaults to commonjs)'
    ).choices(['commonjs', 'esm'])
  );
  outputOptions.option(
    '--clean-output',
    'Clean output directory before generating schemas'
  );
  outputOptions.option(
    '--exclude <regex>',
    'Exclude tables matching this regex'
  );
  outputOptions.option(
    '--include <regex>',
    'Include only tables matching this regex'
  );
  outputOptions.option('--silent', 'Suppress all console output');
  outputOptions.option('--debug', 'Enable debug logging');
  outputOptions.option(
    '--json-schema-import-location <path>',
    'Path to import JSON schemas'
  );
  outputOptions.addOption(
    new Option(
      '--zod-version <value>',
      'Zod version to use. (defaults to 3)'
    ).choices(['3', '4', '4-mini'])
  );
  outputOptions.addOption(
    new Option(
      '--case-transform <true|false>',
      'Whether to do case transformations / conversions for generated schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--singularization <true|false>',
      'Whether to use singularization of type and enum names. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--coerce-dates <true|false>',
      'Whether to use z.coerce.date() for date fields in read schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--stringify-json <true|false>',
      'Whether to JSON.stringify() on json fields in write schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--stringify-dates <true|false>',
      'Whether to convert dates to ISO strings in write schemas. (defaults to false)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--default-empty-array <true|false>',
      'Whether to use empty arrays as defaults for nullable array fields. (defaults to false)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--default-unknown <true|false>',
      'Whether to use "unknown" instead of "any" for unresolved types. (defaults to false)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--default-nulls-to-undefined <true|false>',
      'Whether to transform null values to undefined in generated read schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--object-name-casing <type>',
      'Casing for generated object/type names. (defaults to PascalCase)'
    ).choices(['PascalCase', 'camelCase', 'snake_case'])
  );
  outputOptions.addOption(
    new Option(
      '--field-name-casing <type>',
      'Casing for field/property names in schemas & records. (defaults to camelCase)'
    ).choices(['camelCase', 'snake_case', 'PascalCase', 'passthrough'])
  );

  program.parse();
  const options = program.opts();
  const spinner = createProgressHandler(options.silent);

  try {
    const cliConfig: ZodDbsCliConfig = {
      ...config,
      ...options,
    };

    logDebug('Final configuration:', cliConfig);

    if (!cliConfig.silent) {
      logSettings(cliConfig);
      console.log();
    }

    await generateZodSchemas({
      provider,
      renderer: cliConfig.renderer,
      config: {
        ...cliConfig,
        onProgress: spinner.onProgress,
      },
    });

    spinner.done();
  } catch (error) {
    spinner.fail();

    logError(toError(error).message);
    logDebug(error);

    process.exit(1);
  }
};

const loadProvider = async (
  override: ZodDbsProvider | string | undefined,
  config: ZodDbsCliConfig
) => {
  try {
    const providerOrName =
      override ??
      getArgumentValue('-p') ??
      getArgumentValue('--provider') ??
      config.provider;

    return await importProvider(providerOrName);
  } catch (error) {
    logError(toError(error).message);
    logDebug(error);
    process.exit(1);
  }
};

const addProviderOptions = (program: Command, provider: ZodDbsProvider) => {
  if (!provider.options) return;

  const group = program.optionsGroup(
    `${provider.displayName ?? provider.name} options:`
  );

  for (const option of provider.options) {
    const flag =
      option.type === 'boolean'
        ? `--${option.name}`
        : `--${option.name} <value>`;
    const description = option.description;

    let cliOption = new Option(flag, description);

    if (option.allowedValues) {
      cliOption = cliOption.choices(option.allowedValues);
    }

    if (option.required) {
      cliOption = cliOption.makeOptionMandatory();
    }

    if (option.type === 'number') {
      cliOption = cliOption.argParser((value) => parseInt(value, 10));
    }

    group.addOption(cliOption);
  }
};

const logSettings = (config: ZodDbsCliConfig) => {
  for (const [key, value] of Object.entries(config)) {
    if (value !== undefined && value !== null) {
      logSetting(key, value);
    }
  }
};
