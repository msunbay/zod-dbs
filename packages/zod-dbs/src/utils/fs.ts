import { existsSync, readdirSync, unlinkSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import { logDebug } from 'zod-dbs-core';

import type { ZodDbsTableType } from 'zod-dbs-core';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodDbsTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
