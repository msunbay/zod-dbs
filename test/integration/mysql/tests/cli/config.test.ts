import fs from 'node:fs/promises';
import path from 'node:path';

import { executeCli, getOutputFiles } from '../../../utils/cli.js';
import { getProviderConfig } from '../../../utils/context.js';
import { getOutputDir } from '../../setup.js';

it('CLI generates correct zod schemas with basic options', async () => {
  const outputDir = getOutputDir('config');
  const config = getProviderConfig();

  // Set working directory for config file resolution
  process.env.ZOD_DBS_CWD = path.join(import.meta.dirname, '../../');

  await executeCli(`--port ${config.port}`);

  const outputFiles = await getOutputFiles(outputDir);

  for (const file of outputFiles) {
    const content = await fs.readFile(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
