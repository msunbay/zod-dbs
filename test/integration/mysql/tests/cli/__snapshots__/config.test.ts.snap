// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI generates correct zod schemas with basic options > constants.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export const TABLE_USERS = 'users';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './users/index.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';

export const USERS_ROLES = ['admin', 'editor', 'viewer'] as const;

/**
 * Read schema for the "test.users" table.
 * Validates raw rows read from the database.
 */
export const users_table_schema = z.object({
     /**
      * dataType: int
      */
    id: z.number(),
     /**
      * dataType: varchar
      */
    name: z.string(),
     /**
      * dataType: varchar
      */
    email: z.string().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: varchar
      */
    status: z.string().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: json
      */
    profile: z.json().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: timestamp
      */
    created_at: z.coerce.date().nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: enum
      */
    roles: z.enum(USERS_ROLES).nullable().transform((value) => value ?? undefined).optional(),
});

/**
 * Insert schema for the "test.users" table.
 */
export const users_table_insert_schema = z.object({
     /**
      * dataType: varchar
      */
    name: z.string().max(100),
     /**
      * dataType: varchar
      */
    email: z.string().max(255).nullable().optional(),
     /**
      * dataType: varchar
      * defaultValue: active
      */
    status: z.string().max(20).nullable().optional(),
     /**
      * dataType: json
      */
    profile: z.json().nullable().transform((value) => value ? JSON.stringify(value) : value).optional(),
     /**
      * dataType: timestamp
      * defaultValue: CURRENT_TIMESTAMP
      */
    created_at: z.date().nullable().transform((value) => value ? value.toISOString() : value).optional(),
     /**
      * dataType: enum
      * defaultValue: viewer
      */
    roles: z.enum(USERS_ROLES).nullable().optional(),
});

/**
 * Update schema for the "test.users" table.
 */
export const users_table_update_schema = users_table_insert_schema.partial();

type TableInsertRecord = z.input<typeof users_table_insert_schema>;
type TableReadRecord = z.output<typeof users_table_schema>;

/**
 * Read record for the "test.users" table.
 */
export interface users_record {
    id: TableReadRecord['id'];
    name: TableReadRecord['name'];
    email?: TableReadRecord['email'];
    status?: TableReadRecord['status'];
    profile?: TableReadRecord['profile'];
    created_at?: TableReadRecord['created_at'];
    roles?: TableReadRecord['roles'];
}

/**
 * Insertable record for the "test.users" table.
 */
export interface users_insert_record {
    /**
    * @maxLen: 100
    */
    name: TableInsertRecord['name'];
    /**
    * @maxLen: 255
    */
    email?: TableInsertRecord['email'];
    /**
    * @maxLen: 20
    * @default: active
    */
    status?: TableInsertRecord['status'];
    profile?: TableInsertRecord['profile'];
    /**
    * @default: CURRENT_TIMESTAMP
    */
    created_at?: TableInsertRecord['created_at'];
    /**
    * @maxLen: 6
    * @default: viewer
    */
    roles?: TableInsertRecord['roles'];
}

/**
 * Updatable record for the "test.users" table.
 */
export type users_update_record = Partial<users_insert_record>;

export type users_roles = (typeof USERS_ROLES)[number];
"
`;

exports[`CLI generates correct zod schemas with basic options > types.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export type Table = 
    | 'users'
"
`;
